working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+
+## Requirements
+
+From these sample use cases and others, we can generalize some requirements:
+
+* **Must be able to introspect on the existing program**. Almost all the use
+  cases we want to cover need to be able to look at the fields of an existing
+  class at a minimum, as well as the fields of super classes in many cases. Many
+  also need to inspect the methods, constructors, and parameters of those.
+* **Must be able to synthesize new APIs in existing classes**. Most of the core
+  use cases require adding new methods to classes, whose signatures may not be
+  known ahead of time. Some also require the ability to add new fields, both
+  private and public to classes.
+* **Should be able to synthesize whole new classes**. Some of the core use cases
+  involve creating new classes from either other classes or even just top level
+  functions.
+* **Metaprogramming should be composable**. In one form or another,
+  metaprogramming applications should be able to be composed together. A data
+  class code generator for instance would ideally be a combination of several
+  more targeted code generators for each of the desired features (`copyWith()`,
+  `==`, `hashCode`, etc).
+* **Generated code must be debuggable and user visible**. See the Usability
+  section below.
+
+## Design constraints
+
+In addition to user-level requirements, there are some practical constraints to
+ensure that we don't compromise the existing advantages of Dart.
+
+* **Must support modular compilation**. Any individual library should be able to
+  be fully compiled (including running all macros) using only the transitive
+  dependencies of the library. To this end we should not expose any API or
+  functionality that allows global access to the program, outside of the current
+  library and/or its dependencies. Such functionality would defeat modular
+  compilation.
+* **Must support incremental compilation**. This is key to the hot reload
+  workflow—any change in the program should not have to rerun all
+  metaprogramming in the entire program. It should at worst require a rerun of
+  the metaprogramming in libraries that depend on the modified library.
+* **Must support expression evaluation**. Expression evaluation is a vital tool
+  for the development experience and we need to ensure it is not negatively
+  affected.
+
+For the most part the design constraints here align with the modular compilation
+design constraints. In short, metaprogramming should not introduce global action
+at a distance.
+
+## Non-goals
+
+* **Modifying existing code**. We don’t see a need to allow modifiying existing
+  code through metaprogramming. This leads to unexpected behavior, and
+  negatively affects usability. Dart code that already has clearly defined
+  meaning today should continue to mean the same thing after metaprogramming.
+  The goal is not to be able to hijack existing Dart syntax to do wildly
+  different things, but to augment code with new capabilities.
```

Comment:
One common use-case is to augment existing methods, maybe with an enter/exit protocol, without changing the code that actually runs in the middle. 
So *adding* something to a function, but not *changing* what is there (which really means that the new code can only go before/after/around the original code).

(One possible approach to that, which doesn't change existing code, is to allow a (possibly generated) mixin to be applied *after* the class declarations instead of only *before*. Maybe something like `class Foo with M1 { method(){}  } with M2;` which applies the mixin `M2` on top of `Object&M1{ method(){} }` and makes that the `Foo` class.)

Reply:
Another possible approach is to generate code into a normal mixin, and require that users call `super.<method>` in order to get that behavior (and that they do it in the right place). However I think that is error prone, and is boiler plate of its own.

I do worry a bit about allowing arbitrary new code to be injected at the top/bottom of a method, as that could potentially change the behavior of it, but I do also recognize the usefulness of such an ability. Possibly some way of "wrapping" a function which is sort of what you are describing would make sense.